okhttp/src/main/java/com/squareup/okhttp/OkHttpTlsUpgrader.java

```diff
@@ -69,21 +74,27 @@ public static SSLSocket upgrade(SSLSocketFactory sslSocketFactory,
     spec.apply(sslSocket, getOkHttpRoute(host, port, spec));
 
     Platform platform = Platform.get();
-    try {
-      // Force handshake.
-      sslSocket.startHandshake();
-
-      String negotiatedProtocol = platform.getSelectedProtocol(sslSocket);
-      if (negotiatedProtocol == null) {
-        throw new RuntimeException("protocol negotiation failed");
+    // It's possible that the user provided SSLSocketFactory has already done the handshake
```

Comment:
That was my first thought until I saw https://android.googlesource.com/platform/frameworks/base/+/534a67c/core/java/android/net/SSLCertificateSocketFactory.java

Even it's the case that the socket factory settings won out over our settings, this code should be able to handle it correctly too.